File: game-core/src/main/kotlin/com/thevaguebox/picpac/core/Rules.kt
package com.thevaguebox.picpac.core

object PicPacRules {
    fun newGame(starter: Player = Player.ONE, revision: Long = 0): PicPacState = PicPacState(
        board = Board.EMPTY,
        activePlayer = starter,
        remainingX = 5,
        remainingO = 5,
        phase = PicPacPhase.AwaitingDraw,
        starter = starter,
        revision = revision,
    )

    fun draw(state: PicPacState, symbol: Symbol): TransitionResult<PicPacState> {
        if (state.phase is PicPacPhase.Terminal) return TransitionResult.Rejected(RejectionReason.TERMINAL)
        if (state.phase !is PicPacPhase.AwaitingDraw) return TransitionResult.Rejected(RejectionReason.WRONG_PHASE)
        if ((symbol == Symbol.X && state.remainingX == 0) ||
            (symbol == Symbol.O && state.remainingO == 0)
        ) return TransitionResult.Rejected(RejectionReason.EXHAUSTED_SYMBOL)

        val token = TurnToken(state.revision * 16L + state.board.occupiedCount + 1L)
        val next = state.copy(
            remainingX = state.remainingX - if (symbol == Symbol.X) 1 else 0,
            remainingO = state.remainingO - if (symbol == Symbol.O) 1 else 0,
            phase = PicPacPhase.AwaitingPlacement(symbol, token),
        )
        return TransitionResult.Accepted(next, GameEvent.PieceRevealed(state.activePlayer, symbol))
    }

    fun place(state: PicPacState, cell: Cell, token: TurnToken): TransitionResult<PicPacState> {
        val phase = state.phase
        if (phase is PicPacPhase.Terminal) return TransitionResult.Rejected(RejectionReason.TERMINAL)
        if (phase !is PicPacPhase.AwaitingPlacement) return TransitionResult.Rejected(RejectionReason.WRONG_PHASE)
        if (phase.token != token) return TransitionResult.Rejected(RejectionReason.STALE_TURN)
        if (state.board[cell] != null) return TransitionResult.Rejected(RejectionReason.OCCUPIED)

        val board = state.board.place(cell, phase.held)
        val lines = board.winningLines(phase.held)
        if (lines.isNotEmpty()) {
            val outcome = GameOutcome.Win(state.activePlayer, phase.held, lines)
            return TransitionResult.Accepted(
                state.copy(board = board, phase = PicPacPhase.Terminal(outcome)),
                GameEvent.GameWon(outcome),
            )
        }
        if (board.isFull) {
            return TransitionResult.Accepted(
                state.copy(board = board, phase = PicPacPhase.Terminal(GameOutcome.Draw)),
                GameEvent.GameDrawn,
            )
        }
        return TransitionResult.Accepted(
            state.copy(
                board = board,
                activePlayer = state.activePlayer.other(),
                phase = PicPacPhase.AwaitingDraw,
            ),
            GameEvent.PiecePlaced(state.activePlayer, phase.held, cell),
        )
    }
}

object ClassicRules {
    fun newGame(starter: Player = Player.ONE, revision: Long = 0): ClassicState = ClassicState(
        board = Board.EMPTY,
        activePlayer = starter,
        starter = starter,
        turnToken = token(revision, 0),
        revision = revision,
    )

    fun place(state: ClassicState, cell: Cell, turnToken: TurnToken): TransitionResult<ClassicState> {
        if (state.isTerminal) return TransitionResult.Rejected(RejectionReason.TERMINAL)
        if (state.turnToken != turnToken) return TransitionResult.Rejected(RejectionReason.STALE_TURN)
        if (state.board[cell] != null) return TransitionResult.Rejected(RejectionReason.OCCUPIED)

        val symbol = state.symbolFor(state.activePlayer)
        val board = state.board.place(cell, symbol)
        val lines = board.winningLines(symbol)
        if (lines.isNotEmpty()) {
            val outcome = GameOutcome.Win(state.activePlayer, symbol, lines)
            return TransitionResult.Accepted(
                state.copy(board = board, outcome = outcome),
                GameEvent.GameWon(outcome),
            )
        }
        if (board.isFull) {
            return TransitionResult.Accepted(
                state.copy(board = board, outcome = GameOutcome.Draw),
                GameEvent.GameDrawn,
            )
        }
        val next = state.copy(
            board = board,
            activePlayer = state.activePlayer.other(),
            turnToken = token(state.revision, board.occupiedCount),
        )
        return TransitionResult.Accepted(next, GameEvent.PiecePlaced(state.activePlayer, symbol, cell))
    }

    private fun token(revision: Long, occupied: Int) = TurnToken(revision * 16L + occupied + 1L)
}

File: ios/Packages/PicPacKit/Sources/PicPacCore/Rules.swift
public enum PicPacRules {
    public static func newGame(
        starter: Player = .one,
        revision: Int64 = 0
    ) -> PicPacState {
        precondition((0...TurnToken.maximumRevision).contains(revision))
        // These constants satisfy every domain invariant by construction.
        return try! PicPacState(
            board: .empty,
            activePlayer: starter,
            remainingX: 5,
            remainingO: 5,
            phase: .awaitingDraw,
            starter: starter,
            revision: revision
        )
    }

    public static func draw(
        _ state: PicPacState,
        symbol: Symbol
    ) -> TransitionResult<PicPacState> {
        switch state.phase {
        case .terminal:
            return .rejected(.terminal)
        case .awaitingPlacement:
            return .rejected(.wrongPhase)
        case .awaitingDraw:
            break
        }

        if (symbol == .x && state.remainingX == 0)
            || (symbol == .o && state.remainingO == 0) {
            return .rejected(.exhaustedSymbol)
        }

        let token = token(revision: state.revision, occupiedCount: state.board.occupiedCount)
        let next = try! PicPacState(
            board: state.board,
            activePlayer: state.activePlayer,
            remainingX: state.remainingX - (symbol == .x ? 1 : 0),
            remainingO: state.remainingO - (symbol == .o ? 1 : 0),
            phase: .awaitingPlacement(held: symbol, token: token),
            starter: state.starter,
            revision: state.revision
        )
        return .accepted(
            state: next,
            event: .pieceRevealed(player: state.activePlayer, symbol: symbol)
        )
    }

    public static func place(
        _ state: PicPacState,
        at cell: Cell,
        token: TurnToken
    ) -> TransitionResult<PicPacState> {
        let held: Symbol
        let currentToken: TurnToken
        switch state.phase {
        case .terminal:
            return .rejected(.terminal)
        case .awaitingDraw:
            return .rejected(.wrongPhase)
        case let .awaitingPlacement(symbol, phaseToken):
            held = symbol
            currentToken = phaseToken
        }

        guard currentToken == token else { return .rejected(.staleTurn) }
        guard state.board[cell] == nil else { return .rejected(.occupied) }

        let board = try! state.board.placing(held, at: cell)
        let lines = board.winningLines(for: held)
        if !lines.isEmpty {
            let win = GameOutcome.Win(
                player: state.activePlayer,
                symbol: held,
                lines: lines
            )
            let next = try! PicPacState(
                board: board,
                activePlayer: state.activePlayer,
                remainingX: state.remainingX,
                remainingO: state.remainingO,
                phase: .terminal(outcome: .win(win)),
                starter: state.starter,
                revision: state.revision
            )
            return .accepted(state: next, event: .gameWon(win))
        }

        if board.isFull {
            let next = try! PicPacState(
                board: board,
                activePlayer: state.activePlayer,
                remainingX: state.remainingX,
                remainingO: state.remainingO,
                phase: .terminal(outcome: .draw),
                starter: state.starter,
                revision: state.revision
            )
            return .accepted(state: next, event: .gameDrawn)
        }

        let next = try! PicPacState(
            board: board,
            activePlayer: state.activePlayer.other,
            remainingX: state.remainingX,
            remainingO: state.remainingO,
            phase: .awaitingDraw,
            starter: state.starter,
            revision: state.revision
        )
        return .accepted(
            state: next,
            event: .piecePlaced(player: state.activePlayer, symbol: held, cell: cell)
        )
    }

    private static func token(revision: Int64, occupiedCount: Int) -> TurnToken {
        guard let token = TurnToken.derived(
            revision: revision,
            ordinal: occupiedCount + 1
        ) else {
            preconditionFailure("Revision and occupied count must form a valid turn token")
        }
        return token
    }
}

public enum ClassicRules {
    public static func newGame(
        starter: Player = .one,
        revision: Int64 = 0
    ) -> ClassicState {
        precondition((0...TurnToken.maximumRevision).contains(revision))
        return try! ClassicState(
            board: .empty,
            activePlayer: starter,
            starter: starter,
            turnToken: token(revision: revision, occupiedCount: 0),
            revision: revision
        )
    }

    public static func place(
        _ state: ClassicState,
        at cell: Cell,
        token suppliedToken: TurnToken
    ) -> TransitionResult<ClassicState> {
        guard !state.isTerminal else { return .rejected(.terminal) }
        guard state.turnToken == suppliedToken else { return .rejected(.staleTurn) }
        guard state.board[cell] == nil else { return .rejected(.occupied) }

        let symbol = state.symbol(for: state.activePlayer)
        let board = try! state.board.placing(symbol, at: cell)
        let lines = board.winningLines(for: symbol)
        if !lines.isEmpty {
            let win = GameOutcome.Win(
                player: state.activePlayer,
                symbol: symbol,
                lines: lines
            )
            let next = try! ClassicState(
                board: board,
                activePlayer: state.activePlayer,
                starter: state.starter,
                turnToken: state.turnToken,
                revision: state.revision,
                outcome: .win(win)
            )
            return .accepted(state: next, event: .gameWon(win))
        }

        if board.isFull {
            let next = try! ClassicState(
                board: board,
                activePlayer: state.activePlayer,
                starter: state.starter,
                turnToken: state.turnToken,
                revision: state.revision,
                outcome: .draw
            )
            return .accepted(state: next, event: .gameDrawn)
        }

        let next = try! ClassicState(
            board: board,
            activePlayer: state.activePlayer.other,
            starter: state.starter,
            turnToken: token(revision: state.revision, occupiedCount: board.occupiedCount),
            revision: state.revision
        )
        return .accepted(
            state: next,
            event: .piecePlaced(player: state.activePlayer, symbol: symbol, cell: cell)
        )
    }

    private static func token(revision: Int64, occupiedCount: Int) -> TurnToken {
        guard let token = TurnToken.derived(
            revision: revision,
            ordinal: occupiedCount + 1
        ) else {
            preconditionFailure("Revision and occupied count must form a valid turn token")
        }
        return token
    }
}
